cmd/app/main.go

```diff
@@ -60,11 +66,47 @@ func main() {
 		log.Fatal(err)
 	}
 
-	server := virtual.NewServer(reg, environment)
+	var server Server = virtual.NewServer(reg, environment)
 
 	log.Printf("listening on port: %d\n", *port)
 
-	if err := server.Start(*port); err != nil {
+	go func(server Server) {
+		sig := waitForSignal()
+		log.Printf("received signal: %s", sig.String())
+		shutdown(server, *shutdownTimeout)
+	}(server)
+
+	if err := server.Start(*port); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		log.Printf("received error: %s", err.Error())
+		shutdown(server, *shutdownTimeout)
 		log.Fatal(err)
 	}
 }
+
+type Server interface {
+	Start(int) error
+	Stop(context.Context) error
+}
+
+func waitForSignal() os.Signal {
+	osSig := make(chan os.Signal, 1)
+	signal.Notify(osSig, syscall.SIGTERM)
+	signal.Notify(osSig, syscall.SIGINT)
+
+	// wait for a signal to be received
+	return <-osSig
+}
+
+func shutdown(server Server, timeout time.Duration) {
+	log.Printf("shutting down server with timeout (%s)...", timeout.String())
+	ctx := context.Background()
+	if timeout > 0 { // by default there is no timeout for shutting down
+		ctx, _ = context.WithTimeout(context.Background(), timeout)
```

Comment:
I know it doesn't matter since we're shutting down, but can you just do:

```golang
var cc context.CancelFunc
ctx, cc = context.WithTimeout(context.Background(), timeout)
defer cc()
```